### PR TITLE
Replaces RestSharp library with built-in HttpClientFactory

### LINKED ISSUE #313: MinioClient generates a lot of connections in TIME_WAIT
I ran into some stability problems while listing and downloading metadata from one of our services using Minio. I've traced it down to the RestSharp dependency in the minio-dotnet client. I've added a comment regarding this in an existing issue in the RestSharp repo https://github.com/restsharp/RestSharp/issues/1322

To reproduce issue I've made two equivalent scripts for .net core 2.2 using the most recent (at this time) version of the Minio SDK as well as the Amazon SDK as a reference. You can download them from this gist: https://gist.github.com/DynamicOperationsFacilitator/fc3148bad05552fd5138a12c79399c17 

While listing buckets in an infinite loop isn't really a normal workload. it's a fast operation that highlights the difference in how well persistent HTTP connections work in minio-dotnet vs. awssdk.s3 

Example of how to run:

```
git clone https://gist.github.com/DynamicOperationsFacilitator/fc3148bad05552fd5138a12c79399c17
cd fc3148bad05552fd5138a12c79399c17
docker-compose build && docker-compose up -d
docker-compose exec minioclient netstat -n
docker-compose exec amazons3client netstat -n
```

Sample output from running netstat in each container:

```
$ docker-compose exec minioclient netstat -n |  head
Active Internet connections (w/o servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State
tcp        0      0 172.19.0.4:37547        172.19.0.2:9000         TIME_WAIT
tcp        0      0 172.19.0.4:36883        172.19.0.2:9000         TIME_WAIT
tcp        0      0 172.19.0.4:46841        172.19.0.2:9000         TIME_WAIT
tcp        0      0 172.19.0.4:34095        172.19.0.2:9000         TIME_WAIT
tcp        0      0 172.19.0.4:46111        172.19.0.2:9000         TIME_WAIT
tcp        0      0 172.19.0.4:38251        172.19.0.2:9000         TIME_WAIT
tcp        0      0 172.19.0.4:43549        172.19.0.2:9000         TIME_WAIT
tcp        0      0 172.19.0.4:42931        172.19.0.2:9000         TIME_WAIT

$ docker-compose exec minioclient netstat -n | wc -l
4019
```

```
$ docker-compose exec amazons3client netstat -n
Active Internet connections (w/o servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State
tcp        0    550 172.19.0.5:38061        172.19.0.2:9000         ESTABLISHED
tcp        0      0 172.19.0.5:34379        172.19.0.2:9000         ESTABLISHED
Active UNIX domain sockets (w/o servers)
Proto RefCnt Flags       Type       State         I-Node   Path
```

### PR DESCRIPTION
**Pull request by ebozduman**
Fixes #313 #411
Some of the code came from an old PR, #422, from last year. There are quite number of conflicts to be resolved and the code needed considerable amount of refactoring touches, hence needs a lot of testing.
(The work is still in progress.)

### CODE REVIEW

File: Minio/MinioClient.cs [hunk 1]
```diff
@@ -70,7 +72,10 @@ public partial class MinioClient
         // Enables HTTP tracing if set to true
         private bool trace = false;
 
+        private int requestTimeout;
+
         private const string RegistryAuthHeaderKey = "X-Registry-Auth";
+        private HttpClient HttpClient { get; } = new HttpClient();
```
Comment (chester89) [inline Minio/MinioClient.cs:78]: can I suggest we make this injectable through constructor? for testing/logging purposes

Comment (ebozduman) [reply Minio/MinioClient.cs:78]: Thanks @chester89,
Your suggestion is already part of this PR, #551.

**Comment by harshavardhana** [comment]
@ebozduman do you have an ETA on when do you think we can merge this?

**Comment by ebozduman** [comment]
> @ebozduman do you have an ETA on when do you think we can merge this?

I think I can wrap it up around 2 weeks

**Review by harshavardhana — APPROVED** [review]

File: FileUploader/FileUpload.cs [hunk 2]
```diff
@@ -49,9 +56,10 @@ static void Main()
             {
                 Console.WriteLine(ex.Message);
             }
-            // Added for Windows folks. Without it, the window, tests
-            // run in, dissappears as soon as the test code completes.
-            Console.ReadLine();
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```
Comment (poornas) [inline FileUploader/FileUpload.cs:59]:
```suggestion
            if IsWindows()
```

File: Minio.Examples/Cases/GetBucketReplication.cs [hunk 3]
```diff
@@ -23,7 +23,8 @@ public class GetBucketReplication
     {
         // Get Replication configuration assigned to the bucket
         public async static Task Run(MinioClient minio,
-                                    string bucketName = "my-bucket-name")
+                                    string bucketName = "my-bucket-name",
+                                    string replicationCfgID = "my-replication-ID")
```
Comment (poornas) [inline Minio.Examples/Cases/GetBucketReplication.cs:27]:
```suggestion
                                    string replicationRuleID = "my-replication-ID")
```

File: Minio.Examples/Cases/GetBucketReplication.cs [hunk 4]
```diff
@@ -35,8 +36,15 @@ public class GetBucketReplication
                 if (repl != null && repl.Rules != null && repl.Rules.Count > 0)
                 {
                     Console.WriteLine($"Got Bucket Replication Configuration set for bucket {bucketName}.");
-                    foreach(var rule in repl.Rules)
+                    foreach (var rule in repl.Rules)
                     {
+                        if (rule.ID != replicationCfgID)
+                        {
+                            // failed test due to replication config id mismatch
+                            var errMessage = "Unexpected replication config ID: " +
```
Comment (poornas) [inline Minio.Examples/Cases/GetBucketReplication.cs:44]:
```suggestion
                            var errMessage = "Unexpected replication rule ID: " +
```

File: Minio.Examples/Cases/GetBucketReplication.cs (hunk 3 above)
Comment (poornas) [reply Minio.Examples/Cases/GetBucketReplication.cs:27]: replication config will not have an ID

File: Minio.Examples/Cases/PresignedPostPolicy.cs [hunk 5]
```diff
@@ -21,31 +21,44 @@
...
-                    curlCommand = curlCommand + $" -F {pair.Key}={pair.Value}";
-                }
-                curlCommand = curlCommand + " -F file=@/etc/bashrc " + tuple.Item1; // https://s3.amazonaws.com/my-bucketname";
-                Console.WriteLine(curlCommand);
-            }
-            catch (Exception e)
+            // DefaultPolicy defaultPolicy = delegate (string bucketName,
+            //                                         string objectName,
+            //                                         DateTime expiration)
+            // {
```
Comment (poornas) [inline Minio.Examples/Cases/PresignedPostPolicy.cs:39]: code cleanup

File: Minio.Examples/Cases/PresignedPostPolicy.cs [hunk 6]
```diff
@@ -21,31 +21,44 @@
...
+            PostPolicy form = new PostPolicy();
+            form.SetKey(objectName);
+            form.SetBucket(bucketName);
+            form.SetExpires(expiration);
+            // return policy;
+            // };
+            // PostPolicy postPolicy = defaultPolicy(bucketName,
+            //                                       objectName,
+            //                                       expiration);
```
Comment (poornas) [inline Minio.Examples/Cases/PresignedPostPolicy.cs:48]: remove

File: Minio.Examples/Cases/SetBucketReplication.cs [hunk 7]
```diff
@@ -16,57 +16,116 @@
...
+
+            return result;
+        }
+
         // Set Replication configuration for the bucket
         public async static Task Run(MinioClient minio,
-                                    string bucketName = "my-bucket-name")
+                                    string bucketName = "my-bucket-name",
+                                    string destBucketName = "dest-bucket-name",
+                                    string replicationCfgID = "my-replication-ID")
```
Comment (poornas) [inline Minio.Examples/Cases/SetBucketReplication.cs:54]:
```suggestion
                                    string replicationRuleID = "my-replication-ID")
```

File: Minio.Examples/Cases/GetBucketReplication.cs (hunk 3 above)
Comment (poornas) [inline Minio.Examples/Cases/GetBucketReplication.cs:27]:
```suggestion
                                    string replicationRuleID = "my-replication-ID")
```

**Review by poornas — COMMENTED** [review]
please squash your commits into one

**Comment by harshavardhana** [comment]
@ebozduman please remove any unnecessary commented code - you may or may not squash as we can squash all of them into one commit at the end.

If you do squash all of them you can write a detailed "commit body" that explains what is being done.

File: FileUploader/FileUpload.cs (hunk 2 above)
Comment (ebozduman) [reply FileUploader/FileUpload.cs]: Done

**Review by poornas — CHANGES_REQUESTED** [review]

File: Minio.Examples/Cases/GetBucketReplication.cs (hunk 3 above)
Comment (ebozduman) [reply Minio.Examples/Cases/GetBucketReplication.cs]: Done

File: Minio.Examples/Cases/PresignedPostPolicy.cs (hunk 5 above)
Comment (ebozduman) [reply Minio.Examples/Cases/PresignedPostPolicy.cs:39]: Done.

File: Minio.Examples/Cases/PresignedPostPolicy.cs (hunk 6 above)
Comment (ebozduman) [reply Minio.Examples/Cases/PresignedPostPolicy.cs:48]: Done.

File: Minio.Examples/Cases/SetBucketReplication.cs (hunk 7 above)
Comment (ebozduman) [reply Minio.Examples/Cases/SetBucketReplication.cs]: Done.

File: Minio.Examples/Cases/GetBucketReplication.cs (hunk 3 above)
Comment (ebozduman) [reply Minio.Examples/Cases/GetBucketReplication.cs]: Done.

File: Minio.Functional.Tests/Program.cs [hunk 8]
```diff
@@ -123,7 +129,7 @@ public static void Main(string[] args)
 
                 // Test File GetObject and PutObject functions
                 FunctionalTest.FGetObject_Test1(minioClient).Wait();
-                // FIX=> FPutObject_Test1(minioClient).Wait();
+                // FIX => FPutObject_Test1(minioClient).Wait();
```
[the remark this answers is not in the document]
Comment (ebozduman) [reply Minio.Functional.Tests/Program.cs:132]: Actually, it is part of the testing, see the above line. For some reason there was this commented FIX note here for the same test.
It is just the editor's lint check added a space. 
Shortly, the comment unnecessary. It is removed.

File: Minio.Functional.Tests/Program.cs [hunk 9]
```diff
@@ -160,7 +166,8 @@ public static void Main(string[] args)
                 FunctionalTest.PresignedGetObject_Test2(minioClient).Wait();
                 FunctionalTest.PresignedGetObject_Test3(minioClient).Wait();
                 FunctionalTest.PresignedPutObject_Test1(minioClient).Wait();
-                FunctionalTest.PresignedPutObject_Test2(minioClient).Wait();
+                FunctionalTest.PresignedGetObject_Test1(minioClient).Wait();
+                // FunctionalTest.PresignedPostPolicy_Test1(minioClient).Wait();
```
[the remark this answers is not in the document]
Comment (ebozduman) [reply Minio.Functional.Tests/Program.cs:170]: No. `FunctionalTest.PresignedPutObject_Test2` was mistakenly commented out. 
Removed the comment.

File: Minio/ApiEndpoints/BucketOperations.cs [hunk 10]
```diff
@@ -18,22 +18,58 @@
...
+        public async Task<ListAllMyBucketsResult> ListBucketsAsync(
+            CancellationToken cancellationToken = default(CancellationToken))
+        {
+            // // HttpRequestMessage requestMessage = await this.CreateRequest(HttpMethod.Get, resourcePath: "/").ConfigureAwait(false);
+            // Uri requestUrl = RequestUtil.MakeTargetURL(this.BaseUrl, this.Secure, this.Region);
+            // HttpRequestMessageBuilder requestMessageBuilder = new HttpRequestMessageBuilder(
+            //     HttpMethod.Get, requestUrl);
+            // ResponseResult responseResult = await this.ExecuteTaskAsync(this.NoErrorHandlers, requestMessageBuilder, cancellationToken).ConfigureAwait(false);
+            // ListBucketsResponse listBucketsResponse = new ListBucketsResponse(responseResult.StatusCode, responseResult.Content);
+            // return listBucketsResponse.BucketsResult;
```
[the remark this answers is not in the document]
Comment (ebozduman) [reply Minio/ApiEndpoints/BucketOperations.cs:55]: Done.

File: Minio/ApiEndpoints/ObjectOperations.cs [hunk 11]
```diff
@@ -269,17 +344,54 @@ public async Task<string> PresignedGetObjectAsync(PresignedGetObjectArgs args)
         /// <exception cref="BucketNotFoundException">When bucket is not found</exception>
         /// <exception cref="ObjectNotFoundException">When object is not found</exception>
         /// <exception cref="MalFormedXMLException">When configuration XML provided is invalid</exception>
-        public async Task<Tuple<string, Dictionary<string, string>>> PresignedPostPolicyAsync(PresignedPostPolicyArgs args)
+        public async Task<(Uri, Dictionary<string, string>)> PresignedPostPolicyAsync(PresignedPostPolicyArgs args)
         {
+            // string region = string.Empty;
```
[the remark this answers is not in the document]
Comment (ebozduman) [reply Minio/ApiEndpoints/ObjectOperations.cs:349]: Done.

**Pull request merged**